activerecord/lib/active_record/connection_adapters/mysql2/database_statements.rb

```diff
@@ -48,7 +48,7 @@ def perform_query(raw_connection, sql, binds, type_casted_binds, prepare:, notif
             # made since we established the connection
             raw_connection.query_options[:database_timezone] = default_timezone
 
-            result = if !prepared_statements || binds.nil? || binds.empty?
+            result = if (binds.nil? || binds.empty?) && !prepare
```

Comment:
I wondered about this `!prepare` case... I _think_ the old (7.1) implementation never prepared for empty binds. I note that for Postgres we do prepare in that case (I think I'd misread it earlier as not doing so.)

Given that MySQL [apparently?] always re-plans when executing a prepared statement, when we can't get any benefit from avoiding quoting & extra string manipulation, I wonder whether there's any benefit to doing a prepare, vs the cost of the extra round-trip.

(Happy for us to merge this... mostly just thinking aloud.)

Reply:
Introduced in https://github.com/rails/rails/pull/23461 by @kamipo but no context on whether it's worth using prepared statements if there is no binds.

Reply:
https://dev.mysql.com/doc/refman/8.4/en/sql-prepared-statements.html states that it still saves on parsing the query. Not sure if it amounts to much, but it's something.

Reply:
About #23461, I just ported the old mysql adapter implementation that existed until Rails 4.2.

https://github.com/rails/rails/blob/7847a19f476fb9bee287681586d872ea43785e53/activerecord/lib/active_record/connection_adapters/mysql_adapter.rb#L253-L263

Reply:
💯 It's a bit unfortunate that MySQL doesn't offer a single round trip prepare-and-execute-once-and-destroy protocol command (with the "binary" result set return used by `execute`), but here we are. Using `query` saves two round trips vs. `prepare`, `execute`, `close` though FYI! 👋 I miss ya! 👋

Reply:
👋 ❤️

Reply:
> `close`

Ah true. I wonder if it would make sense for us to default to accumulating "temporary" prepared statements and then bulk flushing them 🤔 

(At minimum it would be a good use case for returning without waiting for the response from the server...)